### DISCUSSION TITLE
Can I pass null as a query key?

### DISCUSSION THREAD

**Discussion by nandorojo** [TanStack/query#2011] [Q&A]
I use SWR and I'm considering switching to React Query. Basically everything about react-query seems better so far, except one: null / error keys.

With SWR, you can do this:

```js
const { data: user } = useUser()
const { data: friends } = useSWR(user ? ['user', user.id] : null, fetcher)
```

With react-query, it seems like you have to just use the `enabled` prop. However, this does not exactly serve the same function. I want to use a nullable key so that I can keep my typescript types strict.

This is how my code usually looks (illustrative)

```ts
const CacheKeys = {
  friends(id: string) {
    type Key = ['user', string]
    const key: Key = ['user', id]

    return key
  }
}

// then in my hook:
const { data: friends } = useSWR(user ? CacheKeys.user(id) : null, fetcher)
```

I want to ensure that my `key` has a strict type. Otherwise, I would have to manually set `enabled: false` and check the conditions of the key. But the conditions can change (it might depend on more than simply `user.id` in the future.)

With the method described above, I can use the `ReturnType` of my `queryKey` in my `queryFn`.

```ts
const Fetchers = {
  friends(...key: ReturnType<typeof CacheKeys['friends']>) {
    return fetchFriends(...)
  }
}
```

It also allows for a `queryKey` to be a function, and if that function throws, the query doesn't fire. This is really useful in many cases as well.

**Comment by TkDodo** [comment]
some time ago, react-query had nullable query keys, but I think with v2 or so, this was changed to the `enabled` option. I haven't had any typescript issues with it, the preferred way is to use inline functions to avoid query key injection and be explicit about the dependencies:

```
const useTodo = (id) => useQuery(['todos', id], () => fetchTodo(id))
```

**Reply by nandorojo** [reply]
The problem with an inline key is that it's harder to revalidate and use the key elsewhere.

**Reply by TkDodo** [reply]
I am personally using queryKeyFactories, colocated with the custom hooks:

```
// features/todos/queries.ts

const todosKey = 'todos'
export const queryKeyFactory = {
    todos: [todosKey],
    todo: (id: number) => [todosKey, id]
}

const fetchTodo = (id: number): Promise<Todos> => axios.get(...)

export const useTodo = (id: number) => useQuery(queryKeyFactory.todo(id), () => fetchTodo(id))
```

**Reply by nandorojo** [reply]
I see. A problem that comes to mind with passing the variables directly to the fetcher (rather than as an argument of the fetcher) is that I might forget a dependency. I suppose I could make the `getTodos(args)` receive the same arguments that the key does. A dependency linter could be useful for that.

My main point is that the current recommendation feels a bit more mistake-prone. I could certainly get used to it, just figured I would mention my current approach.

I'll likely end up building a `useQuery` wrapper that checks for a `null` key and sets `enabled: false` in those cases.

Appreciate your input here by the way, and for letting me think out loud.

**Reply by TkDodo** [reply]
sure, no worries. I think a react-query-eslint-plugin would be awesome to check for missing dependencies. If you are interested, there is also a babel plugin that automatically generates query key dependencies:

https://github.com/dominictwlee/babel-plugin-react-query-key-gen

**Reply by nandorojo** [reply]
I ended up creating a wrapper for `useQuery` that allows null query keys:

```ts
import { useQuery as useReactQuery } from 'react-query'
import type {
  UseQueryOptions,
  QueryKey,
  QueryFunction,
  UseQueryResult,
} from 'react-query'

export function useQuery<
  TQueryFnData = unknown,
  TError = unknown,
  TData = TQueryFnData
>(
  queryKey: QueryKey | null,
  queryFn: QueryFunction<TQueryFnData>,
  {
    enabled = queryKey !== null,
    ...options
  }: UseQueryOptions<TQueryFnData, TError, TData> = {}
): UseQueryResult<TData, TError> {
  return useReactQuery(queryKey, queryFn, { enabled, ...options })
}
```

I prefer the mental model of, if the key is ready, it has all its fields, otherwise it's null and it won't fetch. I'm very used to SWR in this sense.

**Reply by TkDodo** [reply]
sure, there's nothing wrong if you prefer that, and it will work just fine. maybe exclude `enabled` from the `options` - otherwise you could overwrite it from the consumer side ;)

**Reply by nandorojo** [reply]
Right, maybe I'll do that. I figured if I have some specific reason to set `enabled = false` (i.e. if the screen isn't focused), I'll set that, but otherwise if it's undefined it should fallback to the nullable key. I'll keep playing around with that. Thanks @TkDodo!

**Reply by nandorojo** [reply]
<img width="1277" alt="Screen Shot 2021-03-29 at 3 08 40 PM" src="https://user-images.githubusercontent.com/13172299/112887042-b58cc200-90a0-11eb-825e-b7b08460dde0.png">

Hmm, looks like `react-query` doesn't allow this. Is there a work-around? I could pass an empty key, but then different queries would all pass data to the same place. Certainly not intended. I really want to have functions like this:

```ts
const { userId } = useAuth() // userId is null if not signed in

const user = useAccountQuery(userId ? { userId } : null)
```

The only solution was to omit the `enabled` flag, as you mentioned. I then set `queryKey as QueryKey`, since this I assume it's fine to pass `null` when `enabled: false`?

```ts
export function useQuery<
  TQueryFnData = unknown,
  TError = unknown,
  TData = TQueryFnData
>(
  queryKey: QueryKey | null,
  queryFn: QueryFunction<TQueryFnData>,
  options?: Omit<UseQueryOptions<TQueryFnData, TError, TData>, 'enabled'>
): UseQueryResult<TData, TError> {
  return useReactQuery(queryKey as QueryKey, queryFn, {
    enabled: queryKey !== null,
    ...options,
  })
}
```

**Reply by TkDodo** [reply]
yes, the type assertion is necessary - or you just coerce to empty string?

```
useReactQuery(queryKey ?? '')
```

**Reply by nandorojo** [reply]
But if many different queries use an empty string as the key, will they write to the cache? Or does enabled set to false disable any cache interaction?

**Reply by TkDodo** [reply]
if the query is disabled there is no data in the cache (unless you explicitly call refetch on the query). But it's the same as if you pass `null` in and use `null` for the queryKey for multiple queries ;)

the query will show up in the devtools but that's just about it

**Reply by phatmann** [reply]
As stated above, if your query key includes an id and that id can be null (or undefined) because of a dependent query, you have to use the `enabled` flag and then provide some "blank" value for the key or coerce a null key. I have been using `[]` as blank since that feels right, but I still find the code weird to read. Coercing scares me since that null value could cause issues in the React Query code. The worst part of this is that a call to `refetch()` could cause issues, either an error in the query function or bad data being pulled into the cache with the "blank" key.

All of these issues go away if we allow null keys. We won't get a Typescript error, and there is no possibility of poisoning the cache. The semantics are different than `enabled`, since enabled prevents the automatic fetching behavior, while a null key prevents writing to cache. If a key were null, it of course makes sense to also disable automatic fetching, too.

I am not sure why null keys were deprecated in favor of the enabled option. I was unable to find the discussion or PR that explained the decision. Here are the release notes:

> Conditional, falsy or nullish query keys have been replaced by the enabled config flag. This flag must be a boolean-esque value, not a function. If you are looking to replace the throwable function option, you should instead use optional chaining, eg. useQuery(user?.id, fn, { enabled: user?.id }).

Ironically that sample code won't compile under Typescript 😏

**Reply by TkDodo** [reply]
> you have to use the enabled flag and then provide some "blank" value for the key or coerce a null key.

don't understand that. I'm using the enabled prop and never coerce the key to anything, like:

```
const [filter, setFilter] = useState()
useQuery(['something', filter]), () => fetchSomething(filter), { enabled: Boolean(filter) }
```

I don't see how a null key would make that example better?

```
const [filter, setFilter] = useState()
useQuery(filter ? ['something', filter] : null), () => fetchSomething(filter))
```

>  The semantics are different than enabled, since enabled prevents the automatic fetching behavior, while a null key prevents writing to cache.

even for queries with a null key, calling `refetch` would likely still trigger the queryFn. Also, an observer would have to be created in any case (otherwise refetch wouldn't work), so "not writing to the cache" is not true, or at least it depends on how you define writing to the cache. Technically, if you have a disabled query, you get a disabled observer. This is independent from the cache entry, which you may or may not have.

> I am not sure why null keys were deprecated in favor of the enabled option.

I wasn't around when null keys were a thing, but advantages of the enabled option, as far as I can tell, include:
- you can turn queries on an off depending on different conditions. For example, you open a dialog and want to disable the query while the dialog is open. This is easy with `enabled` as it takes any boolean, but kinda weird if you have to manually coerce the key to null if you want that behaviour.
- you can all `setQueryData` on disabled queries and still update the existing cache entry with it.
- observers are likely not permanently disabled. They are often initially disabled, or disabled for some amount of time. They are still "there", and the enabled option reflects that very well.
- You can also have one enabled and one disabled observer of the same query key. This is relevant if you e.g. mount a new component with a disabled query, you won't trigger a refetchOnMount, but you can still get the cached data. Not possible with a null key.

**Reply by nandorojo** [reply]
A null key could make the types better if we passed the filter to the fetcher via its arguments, the way SWR does.

![8CB626D6-7489-46CA-B0C5-4AD58C5C5F33](https://user-images.githubusercontent.com/13172299/135344810-b7d66550-f788-4462-937f-850f6a3f2714.jpeg)

**Reply by phatmann** [reply]
> I don't understand that. I'm using the enabled prop and never coerce the key to anything...

My scenario was not complete. I am using a query keys factory (thanks for this idea, @TkDodo) and one of the methods takes an ID. The ID is derived from another query and can be undefined, yet the query factory method requires a non-null ID. That is where the coercion comes into play.

(A related alternative is to allow the factory method to take a null ID, and return a blank key `[]` in response.)

> ...advantages of the enabled option:
> * you can all `setQueryData` on disabled queries and still update the existing cache entry with it.
> * You can also have one enabled and one disabled observer of the same query key.

Great points. I now see the problem with null keys instead of `enabled`. 

A good solution to the type issue would be to let the query key be a function (or a value). Then you could use the non-null assertion operator without getting an exception:

```
const [filter, setFilter] = useState()
useQuery(() => ['something', filter!], () => fetchSomething(filter), { enabled: !!filter })
```

**Reply by TkDodo** [reply]
the problem is that `enabled: false` only disables the observer, _not_ "the query". If you call `refetch`, the query will still run. I've shown another way of dealing with this apart from the no-null assertion operator here: https://tkdodo.eu/blog/react-query-and-type-script#type-safety-with-the-enabled-option

**Reply by phatmann** [reply]
I keep forgetting about `refetch` even though I mentioned it myself. Grrrrrr.

I have already read all of your blog posts with great interest, they are all excellent. But that particular one does not address how to set up a query factory when IDs can be null. I would like to see your exact solution for that one. Because I am out of ideas that are not unsatisfactory or ineffective in one way or another.

**Reply by TkDodo** [reply]
> But that particular one does not address how to set up a query factory when IDs can be null. I would like to see your exact solution for that one. Because I am out of ideas that are not unsatisfactory or ineffective in one way or another.

I don't really understand the problem, can't we just accept null in the key factory as well?

```
const queryKeys = {
  todos: ["todos"] as const,
  todo: (id: number | null) => [...queryKeys.todos, id] as const
}

const fetchTodo = (id: number | null) => {
  if (!id) {
    return Promise.reject(new Error("invalid id"))
  }
  return axios.get(...)
}

const useTodo = (id: number | null) => useQuery(
  queryKeys.todo(id),
  () => fetchTodo(id),
  { enabled: Boolean(id) }
)
```

**Reply by phatmann** [reply]
Without a valid ID we cannot generate a valid key. By explicitly allowing a null ID, we are making the code harder to reason about just to appease the Typescript compiler. The goal is to convince the compiler that the key will not be accessed unless enabled is true. I have not found a satisfactory implementation of this, but I'm convinced there is one. I'm going to mull it over and hope for a flash of inspiration ⚡️
